working/0731 - horizontal inference/feature-specification.md

```diff
@@ -0,0 +1,485 @@
+   that barring compilation errors, dynamic invocations can only occur if the
+   target has a type of `dynamic` or `Function`._
+
+2. Determine if generic inference is needed.  Generic inference is needed iff
+   the following conditions are all met:
+
+   - There is a target function type.
+
+   - The target function type is generic.
+
+   - The invocation does not explicitly specify type arguments using the
+     _\<typeArguments\>_ syntax.
+
+3. Initial constraints and downwards inference: if generic inference is needed,
+   try to match the return type of the target function type as a subtype of the
+   invocation's type context.  This produces an initial set of type constraints.
+   Then, using those constraints, find the constraint solution for the target
+   function type's type variables, considering bounds.  This produces a
+   preliminary mapping of type variables to type schemas.
+
+4. Visit arguments: Partition the arguments into stages (see [argument
+   partitioning](#Argument-partitioning) below), and then for each stage _k_, do
+   the following:
+
+   * Compute the type context for all invocation arguments in stage _k_.  If
+     generic inference is needed, these contexts are obtained by substituting
+     the preliminary mapping (from the most recent "downwards inference" or
+     "horizontal inference" step) into the corresponding parameter type from the
+     target function type.  Otherwise, if there is a target function type and
+     there are explicit type arguments, they are obtained by substituting the
+     explicit type arguments into the parameter type.  Otherwise, if there is a
+     target function type and no explicit type arguments _(this case can only
+     happen if the target function type is non-generic)_, the parameter type is
+     used directly.  Finally, in the case where this is a dynamic invocation,
+     each type context is `_`.
+
+   * Perform type inference on all invocation arguments in stage _k_ that are
+     not function literal expressions, in source order.  _Note that an invariant
+     of the partitioning is that arguments that are not function literal
+     expressions are always placed in stage zero._
+
+   * Perform type inference on all arguments selected for stage _k_ that are
+     function literal expressions, in source order.  _Note that we do not
+     believe the order of performing type inference on function literal
+     expressions to be user-visible, but we specify source order anyway, to
+     reduce the risk of unpredictable behavior._
+
+   * Constraint generation: if generic inference is needed, try to match the
+     static type of each of the arguments selected for stage _k_ with the
+     corresponding parameter type from the target function type.  This produces
+     additional type constraints beyond those gathered in step 3.
+
+   * Horizontal inference: if generic inference is needed, and this is not the
```

Comment:
This doesn't make clear (per my previous comment) whether we "freeze" the results here or not (presumably we do).  That is, if we finish stage `k` with the constraint `X <: num`, do we finalize `X` as `num`, or do we allow stage `k+1` to further refine the solution for `X` (e.g. by generating a constraint `X <: int`)?  I suspect we want to "freeze" the solution, unfortunately, because otherwise different levels of inference may see different solutions for `X`.  Though maybe that's ok?  It's a little confusing, but then... inference is confusing.  :)

Reply:
Good point.  In https://github.com/dart-lang/language/pull/2236 I've defined precisely when and how my implementation "freezes" the results of type inference.  Hopefully it's clear in that PR that we freeze the type inference of any type parameter as soon as we infer a type schema for it that doesn't contain `_`.